src/auth/impersonated.ts

```diff
@@ -0,0 +1,165 @@
+   * If left unset, sourceCredential must have that role on targetPrincipal.
+   * @param targetScopes scopes to request during the authorization grant.
+   * @param lifetime number of seconds the delegated credential should be
+   * valid for up to 3600 seconds by default, or 43,200 seconds by
+   * extending the token's lifetime, see:
+   * https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oauth
+   * @param endpoint api endpoint override.
+   */
+  constructor(options: ImpersonatedOptions = {}) {
+    super(options);
+    this.credentials = {
+      expiry_date: 1,
+      refresh_token: 'impersonated-placeholder',
+    };
+    this.sourceClient = options.sourceClient || new OAuth2Client();
+    this.targetPrincipal = options.targetPrincipal || '';
+    this.delegates = options.delegates || [];
+    this.targetScopes = options.targetScopes || [];
+    this.lifetime = options.lifetime || 3600;
+    this.endpoint = options.endpoint || 'https://iamcredentials.googleapis.com';
+  }
+
+  /**
+   * Refreshes the access token.
+   * @param refreshToken Unused parameter
+   */
+  protected async refreshToken(
+    refreshToken?: string | null
+  ): Promise<GetTokenResponse> {
+    if (this.isTokenExpiring()) {
+      try {
+        await this.sourceClient.getAccessToken();
+        const name = 'projects/-/serviceAccounts/' + this.targetPrincipal;
+        const u = `${this.endpoint}/v1/${name}:generateAccessToken`;
+        const body = {
+          delegates: this.delegates,
+          scope: this.targetScopes,
+          lifetime: this.lifetime + 's',
+        };
+        const res = await this.sourceClient.request({
+          url: u,
+          data: body,
+          method: 'POST',
+        });
+        const tokenResponse = res.data as TokenResponse;
+        this.credentials.access_token = tokenResponse.accessToken;
+        this.credentials.expiry_date =
+          Date.parse(tokenResponse.expireTime) / 1000;
+        return {
+          tokens: this.credentials,
+          res,
+        };
+      } catch (error) {
+        const status = error?.response?.data?.error?.status;
+        const message = error?.response?.data?.error?.message;
+        if (status && message) {
+          throw new Error(`${status}: unable to impersonate: ${message}`);
+        } else {
+          throw new Error(`unable to impersonate: ${error}`);
+        }
+      }
+    }
+    return {tokens: this.credentials, res: null};
+  }
+
+  protected requestAsync<T>(
+    opts: GaxiosOptions,
+    retry = false
+  ): GaxiosPromise<T> {
+    return super.requestAsync<T>(opts, retry);
+  }
+
+  /**
+   * Get a non-expired access token, after refreshing if necessary.
+   *
+   * @param url The URI being authorized.
+   * @returns An object that includes the authorization header.
+   */
+  async getRequestHeaders(): Promise<Headers> {
+    const res = await this.getRequestMetadataAsync();
```

Comment:
Kinda the same question here as above.  Do we need this to override a base class method?